### PR TITLE
[API Extractor] Groundwork for member trimming

### PR DESCRIPTION
**Pull request by pgonzal**
This PR finally merges a development branch that has been open for several weeks.  It is a major overhaul of the API Extractor semantic analyzer, which in a future PR will enable the the packaging typings generator to trim nested members of definitions.

The package typings generator is still "beta", so the "breaking changes" in this PR don't affect any official features.

NOTE: For any projects with the `packageTypings` feature enabled in their **api-extractor.json**, API Extractor now requires a field like this in your **package.json** file:

```json
  "tsdoc": {
    "tsdocFlavor": "AEDoc"
  },
```

This designation allows API Extractor to safely parse AEDoc comments found in a library dependency, without accidentally parsing comments (and reporting bogus errors) for libraries that were never intended to work with API Extractor.

### CODE REVIEW

File: apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+// See LICENSE in the project root for license information.
+
+import * as ts from 'typescript';
+import { AstSymbol } from './AstSymbol';
+import { Span } from '../../utils/Span';
+
+export interface IAstDeclarationParameters {
+  readonly declaration: ts.Declaration;
+  readonly astSymbol: AstSymbol;
+  readonly parentAstDeclaration: AstDeclaration | undefined;
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts]: >parentAstDeclaration [](start = 11, length = 20)

rename to parent?

File: apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+export class AstDeclaration {
+  public readonly declaration: ts.Declaration;
+
+  public readonly astSymbol: AstSymbol;
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts]: >lic readonly astSymbol: AstSymbol; [](start = 5, length = 34)

comment public members

File: apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  /**
+   * The parent, if this object is nested inside another AstDeclaration.
+   */
+  public readonly parent: AstDeclaration | undefined;
+
+  private readonly _analyzedChildren: AstDeclaration[] = [];
+
+  private readonly _analyzedReferencedAstSymbolsSet: Set<AstSymbol> = new Set<AstSymbol>();
+  private readonly _analyzedReferencedAstSymbols: AstSymbol[] = [];
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts]: >ivate readonly _analyzedReferencedAstSymbols: AstSymbol[] = []; [](start = 4, length = 63)

You might be able to just store the set and convert it to an array when you are ready to iterate

[https://stackoverflow.com/questions/20069828/how-to-convert-set-to-array](https://stackoverflow.com/questions/20069828/how-to-convert-set-to-array)

[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from)

File: apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+   * belonging to its parent, and ignoring references associated with its
+   * children).
+   * @remarks
+   * The collection will be empty until AstSymbol.analyzed is true.
+   */
+  public get referencedAstSymbols(): ReadonlyArray<AstSymbol> {
+    return this.astSymbol.analyzed ? this._analyzedReferencedAstSymbols : [];
+  }
+
+  public notifyChildAttach(child: AstDeclaration): void {
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts]: >notifyChildAttach [](start = 9, length = 17)

jsdoc

File: apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    if (child.parent !== this) {
+      throw new Error('Program Bug: Invalid call to attachChild()');
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts]: >attachChild [](start = 52, length = 11)

fix name

File: apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  /**
+   * Returns a diagnostic dump using Span.getDump(), which reports the detailed
+   * compiler structure.
+   */
+  public getSpanDump(indent: string = ''): string {
+    const span: Span = new Span(this.declaration);
+    return span.getDump(indent);
+  }
+
+  public notifyReferencedAstSymbol(referencedAstSymbol: AstSymbol): void {
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts]: >notifyReferencedAstSymbol [](start = 9, length = 25)

comment

File: apps/api-extractor/src/generators/packageTypings/AstEntryPoint.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { AstSymbol } from './AstSymbol';
+
+export interface IExportedMember {
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstEntryPoint.ts]: >IExportedMember [](start = 17, length = 15)

document this

File: apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+      case ts.SyntaxKind.TypeAliasDeclaration:
+      case ts.SyntaxKind.VariableDeclaration:
+        return true;
+    }
+    return false;
+  }
+
+  public tryGetSymbolForNode(node: ts.Node): ts.Symbol | undefined {
+    return TypeScriptHelpers.getSymbolForDeclaration(node as ts.Declaration);
+  }
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: this is confusing...

File: apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    }
+
+    return this._fetchAstSymbol(symbol, true);
+  }
+
+  private _fetchAstSymbol(symbol: ts.Symbol, addIfMissing: boolean): AstSymbol | undefined {
+    const followAliasesResult: IFollowAliasesResult = SymbolAnalyzer.followAliases(symbol, this._typeChecker);
+
+    const followedSymbol: ts.Symbol = followAliasesResult.followedSymbol;
+    if (followedSymbol.flags & (ts.SymbolFlags.TypeParameter | ts.SymbolFlags.TypeLiteral | ts.SymbolFlags.Transient)) {
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: >ts.SymbolFlags.TypeLiteral [](start = 63, length = 26)

test for type literals?

File: apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+          // so add a mapping for followedSymbol; we'll need it later when renaming identifiers
+          this._astSymbolsBySymbol.set(followedSymbol, astSymbol);
+        }
+      }
+
+      if (!astSymbol) {
+        // None of the above lookups worked, so create a new entry
+        astSymbol = new AstSymbol({
+          localName: followAliasesResult.localName,
+          followedSymbol: followAliasesResult.followedSymbol,
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: >followedSymbol [](start = 10, length = 14)

originalSymbol? "Followed" is a bit weird

File: apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+      }
+    }
+
+    return astSymbol;
+  }
+
+  /**
+   * Returns the first parent satisfying isAstDeclaration(), or undefined if none is found.
+   */
+  private _findFirstParentDeclaration(node: ts.Node): ts.Node | undefined {
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: >D [](start = 26, length = 1)

AstDeclaration

Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: >_findFirstParentDeclaration [](start = 10, length = 27)

tryFind...()

File: apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+          const parentSymbol: ts.Symbol | undefined = this.tryGetSymbolForNode(arbitaryParent);
+          if (!parentSymbol) {
+            throw new Error('Program bug: missing parent symbol for declaration');
+          }
+
+          parentAstSymbol = this._fetchAstSymbol(parentSymbol, addIfMissing);
+        }
+
+        // Okay, now while creating the declarations we will wire them up to the
+        // their corresopnding parent declarations
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: spelling

File: apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+          // If it's an import, add it to the lookup
+          this._entriesByImportKey.set(astImport.key, astSymbol);
+        }
+
+        // We always fetch the entire chain of parents for each declaration.
+        // (Children/siblings are only analyzed on demand.)
+
+        // Is there a parent AstSymbol?
+        const arbitaryParent: ts.Node | undefined
+          = this._findFirstParentDeclaration(followedSymbol.declarations[0]);
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: write down the assumptions

File: apps/api-extractor/src/generators/packageTypings/DtsEntry.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+import { ReleaseTag } from '../../aedoc/ReleaseTag';
+
+export interface IDtsEntryParameters {
+  readonly astSymbol: AstSymbol;
+  readonly originalName: string;
+  readonly exported: boolean;
+  readonly releaseTag: ReleaseTag;
+}
+
+export class DtsEntry {
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/DtsEntry.ts]: >DtsEntry [](start = 13, length = 8)

document what this is for

File: apps/api-extractor/src/generators/packageTypings/DtsEntry.ts [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  public readonly astSymbol: AstSymbol;
+  public readonly originalName: string;
+  public readonly exported: boolean;
+  public readonly releaseTag: ReleaseTag;
+
+  private _nameForEmit: string | undefined = undefined;
+
+  private _sortKey: string|undefined = undefined;
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/DtsEntry.ts]: >g|u [](start = 25, length = 3)

space ?

File: apps/api-extractor/src/generators/packageTypings/SymbolAnalyzer.ts [hunk 16]
```diff
@@ -93,11 +93,30 @@ export class SymbolAnalyzer {
...
 
+    // Is the followedSymbol actually the kind of thing that can be ambient?
+    if (isAmbient) {
+      for (const declaration of current.declarations || []) {
+        switch (declaration.kind) {
+          case ts.SyntaxKind.ClassDeclaration:
+          case ts.SyntaxKind.InterfaceDeclaration:
+          case ts.SyntaxKind.FunctionDeclaration:
+          case ts.SyntaxKind.ModuleDeclaration:
+          case ts.SyntaxKind.VariableDeclaration:
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/SymbolAnalyzer.ts]: this is a subset of the things in the other file

File: apps/api-extractor/src/generators/packageTypings/PackageTypingsGenerator.ts [hunk 17]
```diff
@@ -42,20 +46,60 @@ export enum PackageTypingsDtsKind {
...
    */
   public analyze(): void {
-    const packageSymbol: ts.Symbol = this._context.package.getDeclarationSymbol();
-    this._entryTable.analyze(packageSymbol);
+    if (this._astEntryPoint) {
+      throw new Error('PackageTypingsGenerator.analyze() was already called');
+    }
+
+    // Build the entry point
+    this._astEntryPoint = this._astSymbolTable.fetchEntryPoint(this._context.package.getDeclaration().getSourceFile());
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/PackageTypingsGenerator.ts]: >this._context.package.getDeclaration() [](start = 63, length = 38)

pull into temp variable. this line has a lot going on

File: apps/api-extractor/src/generators/packageTypings/PackageTypingsGenerator.ts [hunk 18]
```diff
@@ -74,14 +118,99 @@ export class PackageTypingsGenerator {
...
+    }
+
+    // Next generate unique names for the non-exports that will be emitted
+    for (const dtsEntry of this._dtsEntries) {
+      if (!dtsEntry.exported) {
+        let suffix: number = 1;
+        dtsEntry.nameForEmit = dtsEntry.originalName;
+
+        while (usedNames.has(dtsEntry.nameForEmit)) {
+          dtsEntry.nameForEmit = dtsEntry.originalName + '_' + ++suffix;
```
Comment (nickpape-msft) [inline apps/api-extractor/src/generators/packageTypings/PackageTypingsGenerator.ts]: >dtsEntry.originalName + '_' + ++suffix [](start = 33, length = 38)

so there aren't a bunch of plusses, use backticks

File: apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts (hunk 3 above)
Comment (pgonzal) [reply apps/api-extractor/src/generators/packageTypings/AstDeclaration.ts]: Great suggestion. I just used the [...set] notation. This works fine now that we're targeting ES6.

---
In reply to: [173999330](https://github.com/Microsoft/web-build-tools/pull/560#discussion_r173999330) [](ancestors = 173999330)

File: apps/api-extractor/src/generators/packageTypings/SymbolAnalyzer.ts (hunk 16 above)
Comment (pgonzal) [reply apps/api-extractor/src/generators/packageTypings/SymbolAnalyzer.ts]: I moved both switch statements into SymbolAnalyzer for now.

---
In reply to: [174002883](https://github.com/Microsoft/web-build-tools/pull/560#discussion_r174002883) [](ancestors = 174002883)

File: apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts (hunk 10 above)
Comment (pgonzal) [reply apps/api-extractor/src/generators/packageTypings/AstSymbolTable.ts]: I'll handle this in a separate PR. I need to think of a better word for "following all the paths to their unique endpoint".  "Dealias" is apparently a legitimate mathematical word, but pretty weird to type.  "Normalize" is sort of the right idea, but I need an unambiguous jargon for this important part of the algorithm.

@nickpape-msft @iclanton

---
In reply to: [174001612](https://github.com/Microsoft/web-build-tools/pull/560#discussion_r174001612) [](ancestors = 174001612)

**Comment by pgonzal** [comment]
:shipit:

**Pull request merged**